libbpf-tools/profile.c

```diff
@@ -0,0 +1,616 @@
+		if (env.cpu != -1 && env.cpu != i)
+			continue;
+
+		fd = syscall(__NR_perf_event_open, &attr, -1, i, -1, 0);
+		if (fd < 0) {
+			/* Ignore CPU that is offline */
+			if (errno == ENODEV)
+				continue;
+			fprintf(stderr, "failed to init perf sampling: %s\n",
+				strerror(errno));
+			return -1;
+		}
+		links[i] = bpf_program__attach_perf_event(prog, fd);
+		if (libbpf_get_error(links[i])) {
+			fprintf(stderr, "failed to attach perf event on cpu: "
+				"%d\n", i);
+			links[i] = NULL;
+			close(fd);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static int stack_id_err(int stack_id)
+{
+	return (stack_id < 0) && (stack_id != -EFAULT);
+}
+
+static int cmp_counts(const void *dx, const void *dy)
+{
+	__u64 x = ((struct key_ext_t *) dx)->v;
+	__u64 y = ((struct key_ext_t *) dy)->v;
+	return x > y ? -1 : !(x == y);
+}
+
+static bool batch_map_ops = true; /* hope for the best */
+
+static bool read_batch_counts_map(int fd, struct key_ext_t *items, __u32 *count)
+{
+	void *in = NULL, *out;
+	__u32 i, n, n_read = 0;
+	int err = 0;
+	__u32 vals[*count];
+	struct key_t keys[*count];
+
+	while (n_read < *count && !err) {
+		n = *count - n_read;
+		err = bpf_map_lookup_batch(fd, &in, &out, keys + n_read,
+					   vals + n_read, &n, NULL);
+		if (err && errno != ENOENT) {
+			/* we want to propagate EINVAL upper, so that
+			 * the batch_map_ops flag is set to false */
+			if (errno != EINVAL)
+				warn("bpf_map_lookup_batch: %s\n",
+				     strerror(-err));
+			return false;
+		}
+		n_read += n;
+		in = out;
+	}
+
+	for (i = 0; i < n_read; i++) {
+		items[i].k.pid = keys[i].pid;
+		items[i].k.kernel_ip = keys[i].kernel_ip;
+		items[i].k.user_stack_id = keys[i].user_stack_id;
+		items[i].k.kern_stack_id = keys[i].kern_stack_id;
+		strncpy(items[i].k.name, keys[i].name, TASK_COMM_LEN);
+		items[i].v = vals[i];
+	}
+
+	*count = n_read;
+	return true;
+}
+
+static bool read_counts_map(int fd, struct key_ext_t *items, __u32 *count)
+{
+	struct key_t empty = {};
+	struct key_t *lookup_key = &empty;
+	int i = 0;
+	int err;
+
+	if (batch_map_ops) {
+		bool ok = read_batch_counts_map(fd, items, count);
+		if (!ok && errno == EINVAL) {
+			/* fall back to a racy variant */
+			batch_map_ops = false;
+		} else {
+			return ok;
+		}
+	}
+
+	if (!items || !count || !*count)
+		return true;
+
+	while (!bpf_map_get_next_key(fd, lookup_key, &items[i].k)) {
+
+		err = bpf_map_lookup_elem(fd, &items[i].k, &items[i].v);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup counts: %d\n", err);
+			return false;
+		}
+		if (items[i].v == 0)
+			continue;
+
+		lookup_key = &items[i].k;
+		i++;
+	}
+
+	*count = i;
+	return true;
+}
+
+static void print_map(struct ksyms *ksyms, struct syms_cache *syms_cache,
```

Comment:
Reduce the nestedness of this function ?

Reply:
I reduced it. Could you please review this?